raiden/transfer/channel.py

```diff
@@ -1213,39 +1183,35 @@ def update_contract_balance(end_state: NettingChannelEndState, contract_balance:
         end_state.contract_balance = contract_balance
 
 
-def compute_merkletree_with(
-    merkletree: MerkleTreeState, lockhash: LockHash
-) -> Optional[MerkleTreeState]:
-    """Register the given lockhash with the existing merkle tree."""
-    # Use None to inform the caller the lockshash is already known
-    result = None
-
-    leaves = merkletree.layers[LEAVES]
-    if lockhash not in leaves:
-        leaves = list(leaves)
-        leaves.append(Keccak256(lockhash))
-        result = MerkleTreeState(compute_layers(leaves))
-
-    return result
-
+def compute_locks_with(
+    locks: PendingLocksState, lock: Union[HashTimeLockState, UnlockPartialProofState]
+) -> Optional[PendingLocksState]:
+    """Register the given lock with as a pending locks."""
+    if bytes(lock.encoded) not in locks.locks:
+        locks = PendingLocksState(list(locks.locks))
+        locks.locks.append(EncodedData(bytes(lock.encoded)))  # pylint: disable=E1101
+        return locks
+    else:
+        return None
 
-def compute_merkletree_without(
-    merkletree: MerkleTreeState, lockhash: LockHash
-) -> Optional[MerkleTreeState]:
-    # Use None to inform the caller the lockhash is unknown
-    result = None
 
-    leaves = merkletree.layers[LEAVES]
-    if lockhash in leaves:
-        leaves = list(leaves)
-        leaves.remove(Keccak256(lockhash))
+def compute_locks_without(
+    locks: PendingLocksState, lock_encoded: EncodedData
+) -> Optional[PendingLocksState]:
+    # Use None to inform the caller the lock is unknown
+    if lock_encoded in locks.locks:
+        locks = PendingLocksState(list(locks.locks))
```

Comment:
Why construct a new state if you already have `locks`?

Reply:
If I don't make a copy, I modify the original locks, and that caused test failures.

Before this PR, this function was already making a copy.  https://github.com/raiden-network/raiden/pull/4192/files#diff-fe7772da6084a3493b31e88fbf7f6d55L1225